packages/lesswrong/lib/sql/Query.ts

```diff
@@ -0,0 +1,509 @@
+   */
+  compile(argOffset = 0, subqueryOffset = 'A'.charCodeAt(0)): {sql: string, args: any[]} {
+    const strings: string[] = [];
+    let args: any[] = [];
+    for (const atom of this.atoms) {
+      if (atom instanceof Arg) {
+        strings.push(`$${++argOffset}${atom.typehint}`);
+        args.push(atom.value);
+      } else if (atom instanceof Query) {
+        strings.push("(");
+        const subquery = this.nameSubqueries ? String.fromCharCode(subqueryOffset++) : "";
+        const result = atom.compile(argOffset, subqueryOffset);
+        strings.push(result.sql);
+        args = args.concat(result.args);
+        argOffset += result.args.length;
+        strings.push(`) ${subquery}`);
+      } else if (atom instanceof Table) {
+        strings.push(`"${atom.getName()}"`);
+      } else {
+        strings.push(atom);
+      }
+    }
+    return {
+      sql: strings.join(" "),
+      args,
+    };
+  }
+
+  /**
+   * Lookup a field in the current scope.
+   */
+  getField(name: string): Type | undefined {
+    return this.getFields()[name] ?? this.table?.getField(name);
+  }
+
+  /**
+   * Get all the fields defined in the current scope (does not return fields defined
+   * in parent scopes).
+   */
+  getFields(): Record<string, Type> {
+    return this.table instanceof Query ? this.table.syntheticFields : this.table.getFields();
+  }
+
+  /**
+   * Internal helper to create a new Arg - allows us to encapsulate Arg
+   * locally in this file.
+   */
+  protected createArg(value: any, type?: Type) {
+    return new Arg(value, type);
+  }
+
+  /**
+   * In complex queries, we (very) often need to provide typehints to make Postgres
+   * happy (especially when using aggregations which result in subqueries with
+   * synthetic fields). `getTypeHint` attempts to provide such a hint, returning the
+   * empty string if one cannot be determined.
+   *
+   * If `typeHint` is an instance of `Type` then that type will be used, otherwise
+   * `typeHint` is assumed to be the value we are getting the type for.
+   */
+  private getTypeHint(typeHint?: any): string {
+    if (typeHint instanceof Type) {
+      return "::" + typeHint.toConcrete().toString();
+    }
+    switch (typeof typeHint) {
+      case "number":
+        return Number.isInteger(typeHint) ? "::INTEGER" : "::REAL";
+      case "string":
+        return "::TEXT";
+      case "boolean":
+        return "::BOOL";
+    }
+    if (typeHint instanceof Date) {
+      return "::TIMESTAMPTZ";
+    }
+    return "";
+  }
+
+  /**
+   * Sometimes, we require two values to have the same type hint (for instance, when
+   * using binary comparison operators), even if those values could be given more
+   * precise type hints separately in isolation.
+   */
+  private getUnifiedTypeHint(a: Atom<T>[], b: Atom<T>[]): string | undefined {
+    const aArg = a.find((atom) => atom instanceof Arg) as Arg;
+    const bArg = b.find((atom) => atom instanceof Arg) as Arg;
```

Comment:
These can use a type guard instead of a cast: `a.find((atom): atom is Arg => atom instanceof Arg);`.  This also preserves the possibility that they're undefined :)